ci/github-script/merge: improve merge operation and error messages

wolfgangwalther opened the pull request:
We previously used auto-merge first and then enqueued explicitly on the assumption that auto-merge would fail if the PR was actually in mergeable state already. This turned out to be false.

Instead, we currently face the problem of auto-merge sometimes getting stuck. This seems to happen when, at the time of enabling auto-merge, the required status checks already passed and the PR would be ready to go - but sometimes GitHub doesn't do it. This *can* be unblocked by approving the PR again, which seems to run the internal "let's check whether we can merge this" procedures on the GitHub side again.

However, we can probably also solve this by just explicitly trying to enqueue the PR first - and only if that fails, fall back to auto-merge. I previously argued against that, based on a potential race condition, in which a PR could become ready to merge between these two requests - at which point the auto-merge operation would fail, if the original assumption was true. But since we don't observe this, we might as well switch.

---

When a user tries to merge a PR, but is not allowed to, it is helpful to explicitly list the users who *are* allowed. This helps explaining *why* the merge-eligible label was set.

I objected to this proposal before, because it would incur too many API requests. But after we have restructured the checklist, this is not actually true anymore - we can now sensibly run this only when a comment is posted and not whenever we check a PR for eligibility.

## Things done

- [x] Tested locally.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
ci/github-script/merge: switch order of merge operations

We previously used auto-merge first and then enqueued explicitly on the
assumption that auto-merge would fail if the PR was actually in
mergeable state already. This turned out to be false.

Instead, we currently face the problem of auto-merge sometimes getting
stuck. This seems to happen when, at the time of enabling auto-merge,
the required status checks already passed and the PR would be ready to
go - but sometimes GitHub doesn't do it. This *can* be unblocked by
approving the PR again, which seems to run the internal "let's check
whether we can merge this" procedures on the GitHub side again.

However, we can probably also solve this by just explicitly trying to
enqueue the PR first - and only if that fails, fall back to auto-merge.
I previously argued against that, based on a potential race condition,
in which a PR could become ready to merge between these two requests -
at which point the auto-merge operation would fail, if the original
assumption was true. But since we don't observe this, we might as well
switch.

wolfgangwalther committed:
ci/github-script/merge: list eligible users in comment

When a user tries to merge a PR, but is not allowed to, it is helpful to
explicitly list the users who *are* allowed. This helps explaining *why*
the merge-eligible label was set.

I objected to this proposal before, because it would incur too many API
requests. But after we have restructured the checklist, this is not
actually true anymore - we can now sensibly run this only when a comment
is posted and not whenever we check a PR for eligibility.

wolfgangwalther committed:
ci/github-script/bot: move getTeamMembers cache into main file

This allows re-using this elsewhere with a shared cache.

MattSturgeon commented on ci/github-script/merge.js:
```diff
@@ -291,6 +292,22 @@ async function handleMerge({
       '',
     ]
 
+    if (!eligible.has(comment.user.id)) {
```
Do we also need to null-check `user` : `comment.user?.id` or `comment.user && !eligible.has(comment.user.id)`

e.g. if the comment was posted and then the user deleted before we handle it.

MattSturgeon reviewed (commented):
Nice

wolfgangwalther committed:
ci/github-script/merge: improve testability

By only ignoring already-handled comments when running non-dry, it's
much easier to look at existing PRs, for which the merge bot already
commented, and iterate on them locally.

It's dry mode anyway, so it won't hurt to get a few more merge comments
in the console output.

wolfgangwalther commented:
I'm confused why CI fails on this one. This works perfectly well for me locally.

MattSturgeon commented on ci/github-script/bot.js:
```diff
@@ -96,6 +96,41 @@ module.exports = async ({ github, context, core, dry }) => {
+  const users = {}
+  function getUser(id) {
+    if (!users[id]) {
+      users[id] = github.request(`GET /user/${id}`).then((resp) => resp.data)
```
Looking at https://github.com/octokit/request.js#rest-api-example

It seems the intended usage is:

```suggestion
      users[id] = github.request("GET /user/{id}", {id}).then((resp) => resp.data)
```

MattSturgeon commented on ci/github-script/merge.js:
```diff
@@ -295,6 +272,18 @@ async function handleMerge({
 
+    if (eligible.size > 0 && !eligible.has(comment.user.id)) {
+      const users = await Promise.all(
+        Array.from(eligible).map(async (id) => (await getUser(id)).login),
```
We can do this (I think):

```suggestion
        Array.from(eligible, async (id) => (await getUser(id)).login),
```
